File: 02两数相加.cpp
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *virtualHead = new ListNode(-1);
        ListNode *k = virtualHead;
        int flag = 0;
        int t;
        while (l1 || l2)
        {
            int n1 = (l1 ? l1->val : 0);
            int n2 = (l2 ? l2->val : 0);
            t = n1 + n2 + flag;
            flag = t > 9 ? 1 : 0;
            k->next = new ListNode(t % 10);
            k = k->next;
            l1 = (l1 ? l1->next : l1);
            l2 = (l2 ? l2->next : l2);
        }
        if (flag)
        {
            k->next = new ListNode(1);
        }
        return virtualHead->next;
    }
};

File: 02两数相加.py
from typing import *


# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def addTwoNumbers(self, l1: Optional[ListNode],
                      l2: Optional[ListNode]) -> Optional[ListNode]:
        virtualHead = ListNode(-1)
        k = virtualHead
        flag = 0
        while l1 or l2:
            n1 = l1.val if l1 is not None else 0
            n2 = l2.val if l2 is not None else 0
            n = n1 + n2 + flag
            k.next = ListNode(n % 10)
            flag = 1 if n > 9 else 0
            k = k.next
            l1 = l1.next if l1 is not None else l1
            l2 = l2.next if l2 is not None else l2
        if flag:
            k.next = ListNode(1)
        return virtualHead.next
